File: Qt6_cpp/pyramid_scroll.cpp
#include <QGuiApplication>
#include <Qt3DCore/QEntity>
#include <Qt3DExtras/Qt3DWindow>
#include <Qt3DRender/QCamera>
#include <Qt3DExtras/QForwardRenderer>
#include <Qt3DExtras/QPhongMaterial>
#include <Qt3DExtras/QOrbitCameraController>
#include <Qt3DCore/QTransform>
#include <Qt3DRender/QGeometryRenderer>
#include <Qt3DCore/QGeometry>
#include <Qt3DCore/QBuffer>
#include <Qt3DCore/QAttribute>
#include <Qt3DRender/QObjectPicker>
#include <Qt3DRender/QPickEvent>
#include <Qt3DInput/QInputSettings>
#include <QColor>
#include <QMouseEvent>
#include <QWheelEvent>
#include <array>
#include <vector>
#include <iostream>

// Structure to hold edge line entities
struct EdgeLines {
    std::vector<Qt3DCore::QEntity*> edges;
    std::vector<Qt3DExtras::QPhongMaterial*> materials;
    int selectedEdge = -1;
};

Qt3DCore::QEntity* createEdgeLine(Qt3DCore::QEntity* parent, 
                                   const QVector3D& start, 
                                   const QVector3D& end,
                                   const QColor& color)
{
    using namespace Qt3DRender;
    using namespace Qt3DCore;

    std::array<float, 6> vertices = {
        start.x(), start.y(), start.z(),
        end.x(), end.y(), end.z()
    };

    QGeometry* geometry = new QGeometry(parent);

    QBuffer* vertexDataBuffer = new QBuffer(geometry);
    vertexDataBuffer->setData(QByteArray(reinterpret_cast<const char*>(vertices.data()),
                                         vertices.size() * sizeof(float)));

    QAttribute* positionAttribute = new QAttribute();
    positionAttribute->setName(QAttribute::defaultPositionAttributeName());
    positionAttribute->setVertexBaseType(QAttribute::Float);
    positionAttribute->setVertexSize(3);
    positionAttribute->setAttributeType(QAttribute::VertexAttribute);
    positionAttribute->setBuffer(vertexDataBuffer);
    positionAttribute->setByteStride(3 * sizeof(float));
    positionAttribute->setCount(2);
    geometry->addAttribute(positionAttribute);

    QGeometryRenderer* lineRenderer = new QGeometryRenderer();
    lineRenderer->setGeometry(geometry);
    lineRenderer->setPrimitiveType(QGeometryRenderer::Lines);

    Qt3DExtras::QPhongMaterial* material = new Qt3DExtras::QPhongMaterial();
    material->setDiffuse(color);
    material->setAmbient(color);

    QEntity* lineEntity = new QEntity(parent);
    lineEntity->addComponent(lineRenderer);
    lineEntity->addComponent(material);

    return lineEntity;
}

EdgeLines* createPyramidWithEdges(Qt3DCore::QEntity* parent, Qt3DCore::QTransform* sharedTransform)
{
    using namespace Qt3DRender;
    using namespace Qt3DCore;

    // Container entity
    QEntity* pyramidContainer = new QEntity(parent);

    // Vertices of a pyramid (tip + base)
    std::array<float, 15> vertices = {
         0.0f,  0.5f,  0.0f,  // 0: tip
        -0.5f, -0.5f,  0.5f,  // 1: base front-left
         0.5f, -0.5f,  0.5f,  // 2: base front-right
         0.5f, -0.5f, -0.5f,  // 3: base back-right
        -0.5f, -0.5f, -0.5f   // 4: base back-left
    };

    // Triangle indices (4 sides + base as 2 triangles)
    std::array<unsigned int, 18> indices = {
        0,1,2,  // front
        0,2,3,  // right
        0,3,4,  // back
        0,4,1,  // left
        1,2,4,  // base triangle 1
        2,3,4   // base triangle 2
    };

    QGeometry* geometry = new QGeometry(pyramidContainer);

    // Vertex buffer
    QBuffer* vertexDataBuffer = new QBuffer(geometry);
    vertexDataBuffer->setData(QByteArray(reinterpret_cast<const char*>(vertices.data()),
                                         vertices.size() * sizeof(float)));

    QAttribute* positionAttribute = new QAttribute();
    positionAttribute->setName(QAttribute::defaultPositionAttributeName());
    positionAttribute->setVertexBaseType(QAttribute::Float);
    positionAttribute->setVertexSize(3);
    positionAttribute->setAttributeType(QAttribute::VertexAttribute);
    positionAttribute->setBuffer(vertexDataBuffer);
    positionAttribute->setByteStride(3 * sizeof(float));
    positionAttribute->setCount(vertices.size()/3);
    geometry->addAttribute(positionAttribute);

    // Index buffer
    QBuffer* indexDataBuffer = new QBuffer(geometry);
    indexDataBuffer->setData(QByteArray(reinterpret_cast<const char*>(indices.data()),
                                        indices.size() * sizeof(unsigned int)));

    QAttribute* indexAttribute = new QAttribute();
    indexAttribute->setAttributeType(QAttribute::IndexAttribute);
    indexAttribute->setVertexBaseType(QAttribute::UnsignedInt);
    indexAttribute->setBuffer(indexDataBuffer);
    indexAttribute->setCount(indices.size());
    geometry->addAttribute(indexAttribute);

    // Geometry renderer
    QGeometryRenderer* pyramidRenderer = new QGeometryRenderer();
    pyramidRenderer->setGeometry(geometry);
    pyramidRenderer->setPrimitiveType(QGeometryRenderer::Triangles);

    // Material
    Qt3DExtras::QPhongMaterial* material = new Qt3DExtras::QPhongMaterial();
    material->setDiffuse(Qt::white);

    // Pyramid entity
    QEntity* pyramidEntity = new QEntity(pyramidContainer);
    pyramidEntity->addComponent(pyramidRenderer);
    pyramidEntity->addComponent(material);
    pyramidEntity->addComponent(sharedTransform);

    // Create edge lines
    EdgeLines* edgeLines = new EdgeLines();
    
    // Define pyramid edges (vertex pairs)
    std::vector<std::pair<int, int>> edgePairs = {
        {0, 1}, {0, 2}, {0, 3}, {0, 4},  // tip to base corners
        {1, 2}, {2, 3}, {3, 4}, {4, 1}   // base edges
    };

    // Convert vertex indices to 3D points
    auto getVertex = [&vertices](int idx) {
        return QVector3D(vertices[idx*3], vertices[idx*3+1], vertices[idx*3+2]);
    };

    // Create edge line entities
    for (const auto& edge : edgePairs) {
        QEntity* edgeEntity = createEdgeLine(pyramidContainer, 
                                             getVertex(edge.first), 
                                             getVertex(edge.second),
                                             Qt::yellow);
        edgeEntity->addComponent(sharedTransform);
        
        // Add picker to edge
        Qt3DRender::QObjectPicker* picker = new Qt3DRender::QObjectPicker(edgeEntity);
        picker->setEnabled(true);
        edgeEntity->addComponent(picker);
        
        edgeLines->edges.push_back(edgeEntity);
        
        // Store material for later color changes
        auto components = edgeEntity->components();
        for (auto* comp : components) {
            if (auto* mat = qobject_cast<Qt3DExtras::QPhongMaterial*>(comp)) {
                edgeLines->materials.push_back(mat);
                break;
            }
        }
    }

    return edgeLines;
}

// Custom window class to capture mouse events
class CustomQt3DWindow : public Qt3DExtras::Qt3DWindow
{
protected:
    void mousePressEvent(QMouseEvent* event) override
    {
        std::cout << "Mouse Press: button=" << event->button() 
                  << " pos=(" << event->pos().x() << "," << event->pos().y() << ")" << std::endl;
        Qt3DExtras::Qt3DWindow::mousePressEvent(event);
    }
    
    void mouseReleaseEvent(QMouseEvent* event) override
    {
        std::cout << "Mouse Release: button=" << event->button() << std::endl;
        Qt3DExtras::Qt3DWindow::mouseReleaseEvent(event);
    }
    
    void mouseMoveEvent(QMouseEvent* event) override
    {
        std::cout << "Mouse Move: pos=(" << event->pos().x() << "," << event->pos().y() << ")" << std::endl;
        Qt3DExtras::Qt3DWindow::mouseMoveEvent(event);
    }
    
    void wheelEvent(QWheelEvent* event) override
    {
        std::cout << "Wheel Event: angleDelta=(" << event->angleDelta().x() 
                  << "," << event->angleDelta().y() << ")" << std::endl;
        Qt3DExtras::Qt3DWindow::wheelEvent(event);
    }
};

int main(int argc, char** argv)
{
    QGuiApplication app(argc, argv);

    // 3D Window with mouse event logging
    CustomQt3DWindow* view = new CustomQt3DWindow();

    // Set clear color
    Qt3DExtras::QForwardRenderer* frameGraph =
        qobject_cast<Qt3DExtras::QForwardRenderer*>(view->defaultFrameGraph());
    if(frameGraph)
        frameGraph->setClearColor(QColor(204, 51, 128)); // pink

    view->resize(600, 600);
    view->show();

    // Root entity
    Qt3DCore::QEntity* rootEntity = new Qt3DCore::QEntity();

    // Shared transform for rotation
    Qt3DCore::QTransform* transform = new Qt3DCore::QTransform();

    // Add pyramid with edges
    EdgeLines* edgeLines = createPyramidWithEdges(rootEntity, transform);

    // Setup edge picking
    for (size_t i = 0; i < edgeLines->edges.size(); ++i) {
        Qt3DCore::QEntity* edge = edgeLines->edges[i];
        auto components = edge->components();
        
        for (auto* comp : components) {
            if (auto* picker = qobject_cast<Qt3DRender::QObjectPicker*>(comp)) {
                QObject::connect(picker, &Qt3DRender::QObjectPicker::clicked, 
                    [edgeLines, i]() {
                        // Deselect previous edge
                        if (edgeLines->selectedEdge >= 0) {
                            edgeLines->materials[edgeLines->selectedEdge]->setDiffuse(Qt::yellow);
                            edgeLines->materials[edgeLines->selectedEdge]->setAmbient(Qt::yellow);
                        }
                        
                        // Select new edge
                        if (edgeLines->selectedEdge == static_cast<int>(i)) {
                            // Toggle off if clicking same edge
                            edgeLines->selectedEdge = -1;
                        } else {
                            edgeLines->selectedEdge = static_cast<int>(i);
                            edgeLines->materials[i]->setDiffuse(Qt::green);
                            edgeLines->materials[i]->setAmbient(Qt::green);
                            std::cout << "Selected edge " << i << std::endl;
                        }
                    });
                break;
            }
        }
    }

    // Camera
    Qt3DRender::QCamera* camera = view->camera();
    camera->lens()->setPerspectiveProjection(45.0f, 1.0f, 0.1f, 1000.0f);
    camera->setPosition(QVector3D(0, 1.0f, 3.0f));
    camera->setViewCenter(QVector3D(0, 0, 0));

    // Set root entity first
    view->setRootEntity(rootEntity);

    // Orbit camera controller for rotation with middle mouse button
    // Must be created AFTER setting root entity
    Qt3DExtras::QOrbitCameraController* cameraController = 
        new Qt3DExtras::QOrbitCameraController(rootEntity);
    cameraController->setCamera(camera);
    cameraController->setLinearSpeed(50.0f);  // Enable zoom with scroll wheel
    cameraController->setLookSpeed(180.0f);  // Control rotation sensitivity

    return app.exec();
}

File: Pyside6/pyramid_scroll.py
import sys
from array import array
from PySide6.QtCore import Qt, QObject, Slot, QByteArray
from PySide6.QtGui import QGuiApplication, QVector3D, QColor, QMouseEvent, QWheelEvent
from PySide6.Qt3DCore import Qt3DCore
from PySide6.Qt3DRender import Qt3DRender
from PySide6.Qt3DExtras import Qt3DExtras
from PySide6.Qt3DInput import Qt3DInput

class EdgeLines:
    """Replaces the C++ struct for managing edge states."""
    def __init__(self):
        self.edges = []
        self.materials = []
        self.selected_edge = -1

class CustomQt3DWindow(Qt3DExtras.Qt3DWindow):
    """Subclass to log mouse events as per the C++ code."""
    def mousePressEvent(self, event: QMouseEvent):
        print(f"Mouse Press: button={event.button()} pos=({event.position().x()},{event.position().y()})")
        super().mousePressEvent(event)

    def mouseReleaseEvent(self, event: QMouseEvent):
        print(f"Mouse Release: button={event.button()}")
        super().mouseReleaseEvent(event)

    def mouseMoveEvent(self, event: QMouseEvent):
        # Note: Mouse move logs can be very chatty
        super().mouseMoveEvent(event)

    def wheelEvent(self, event: QWheelEvent):
        print(f"Wheel Event: angleDelta=({event.angleDelta().x()},{event.angleDelta().y()})")
        super().wheelEvent(event)

def create_edge_line(parent, start, end, color):
    # Vertex data for a single line (2 points)
    vertices = array("f", [
        start.x(), start.y(), start.z(),
        end.x(), end.y(), end.z()
    ])

    geometry = Qt3DCore.QGeometry(parent)
    
    #buf = Qt3DCore.QBuffer(geometry)
    #buf.setData(vertices.tobytes())

    # Use QByteArray
    buf = Qt3DCore.QBuffer(geometry)
    v_data = QByteArray(vertices.tobytes())
    buf.setData(v_data)

    attr = Qt3DCore.QAttribute(geometry)
    attr.setName(Qt3DCore.QAttribute.defaultPositionAttributeName())
    attr.setVertexBaseType(Qt3DCore.QAttribute.Float)
    attr.setVertexSize(3)
    attr.setAttributeType(Qt3DCore.QAttribute.VertexAttribute)
    attr.setBuffer(buf)
    attr.setByteStride(3 * 4)
    attr.setCount(2)
    geometry.addAttribute(attr)

    renderer = Qt3DRender.QGeometryRenderer(parent)
    renderer.setGeometry(geometry)
    renderer.setPrimitiveType(Qt3DRender.QGeometryRenderer.Lines)

    material = Qt3DExtras.QPhongMaterial(parent)
    material.setDiffuse(color)
    material.setAmbient(color)

    entity = Qt3DCore.QEntity(parent)
    entity.addComponent(renderer)
    entity.addComponent(material)
    return entity

def create_pyramid_with_edges(parent, shared_transform):
    container = Qt3DCore.QEntity(parent)

    # Pyramid data
    vertices = array("f", [
         0.0,  0.5,  0.0,  # 0: tip
        -0.5, -0.5,  0.5,  # 1: base FL
         0.5, -0.5,  0.5,  # 2: base FR
         0.5, -0.5, -0.5,  # 3: base BR
        -0.5, -0.5, -0.5   # 4: base BL
    ])
    indices = array("I", [
        0, 1, 2,  0, 2, 3,  0, 3, 4,  0, 4, 1, # sides
        1, 2, 4,  2, 3, 4                    # base
    ])

    geometry = Qt3DCore.QGeometry(container)

    # Vertex Buffer
    v_buf = Qt3DCore.QBuffer(geometry)
    v_buf.setData(QByteArray(vertices.tobytes()))
    v_attr = Qt3DCore.QAttribute(geometry)
    v_attr.setName(Qt3DCore.QAttribute.defaultPositionAttributeName())
    v_attr.setVertexBaseType(Qt3DCore.QAttribute.Float)
    v_attr.setVertexSize(3)
    v_attr.setAttributeType(Qt3DCore.QAttribute.VertexAttribute)
    v_attr.setBuffer(v_buf)
    v_attr.setCount(5)
    geometry.addAttribute(v_attr)

    # Index Buffer
    i_buf = Qt3DCore.QBuffer(geometry)
    i_buf.setData(QByteArray(indices.tobytes()))
    i_attr = Qt3DCore.QAttribute(geometry)
    i_attr.setVertexBaseType(Qt3DCore.QAttribute.UnsignedInt)
    i_attr.setAttributeType(Qt3DCore.QAttribute.IndexAttribute)
    i_attr.setBuffer(i_buf)
    i_attr.setCount(len(indices))
    geometry.addAttribute(i_attr)

    pyramid_renderer = Qt3DRender.QGeometryRenderer(container)
    pyramid_renderer.setGeometry(geometry)
    pyramid_renderer.setPrimitiveType(Qt3DRender.QGeometryRenderer.Triangles)

    pyramid_mat = Qt3DExtras.QPhongMaterial(container)
    pyramid_mat.setDiffuse(QColor(Qt.white))
    pyramid_mat.setAmbient(QColor(30, 30, 30))

    pyramid_entity = Qt3DCore.QEntity(container)
    pyramid_entity.addComponent(pyramid_renderer)
    pyramid_entity.addComponent(pyramid_mat)
    pyramid_entity.addComponent(shared_transform)

    # Logic for edge creation
    edge_lines = EdgeLines()
    edge_pairs = [
        (0, 1), (0, 2), (0, 3), (0, 4),
        (1, 2), (2, 3), (3, 4), (4, 1)
    ]

    def get_v3d(idx):
        return QVector3D(vertices[idx*3], vertices[idx*3+1], vertices[idx*3+2])

    for i, (v1, v2) in enumerate(edge_pairs):
        edge_ent = create_edge_line(container, get_v3d(v1), get_v3d(v2), QColor(Qt.yellow))
        edge_ent.addComponent(shared_transform)
        
        picker = Qt3DRender.QObjectPicker(edge_ent)
        picker.setEnabled(True)
        edge_ent.addComponent(picker)
        
        edge_lines.edges.append(edge_ent)
        
        # In Python, we can find the material easily
        for comp in edge_ent.components():
            if isinstance(comp, Qt3DExtras.QPhongMaterial):
                edge_lines.materials.append(comp)
                break
                
    return edge_lines

def main():
    app = QGuiApplication(sys.argv)
    view = CustomQt3DWindow()

    # FrameGraph setup for clear color (NVIDIA driver friendly)
    frame_graph = view.defaultFrameGraph()
    if isinstance(frame_graph, Qt3DExtras.QForwardRenderer):
        frame_graph.setClearColor(QColor(204, 51, 128))

    view.resize(800, 800)
    view.show()

    root_entity = Qt3DCore.QEntity()

    # Light
    light_entity = Qt3DCore.QEntity(root_entity)
    light = Qt3DRender.QPointLight(light_entity)
    light.setColor(QColor("white"))
    light.setIntensity(1.0)
    light_entity.addComponent(light)
    light_transform = Qt3DCore.QTransform()
    light_transform.setTranslation(QVector3D(5, 5, 10))
    light_entity.addComponent(light_transform)

    # Input settings
    input_settings = Qt3DInput.QInputSettings(root_entity)

    transform = Qt3DCore.QTransform()
    edge_lines = create_pyramid_with_edges(root_entity, transform)

    # Edge Picking Logic
    def handle_click(index):
        # Deselect old
        if edge_lines.selected_edge >= 0:
            old_mat = edge_lines.materials[edge_lines.selected_edge]
            old_mat.setDiffuse(QColor(Qt.yellow))
            old_mat.setAmbient(QColor(Qt.yellow))
        
        # Select new (or toggle off)
        if edge_lines.selected_edge == index:
            edge_lines.selected_edge = -1
        else:
            edge_lines.selected_edge = index
            new_mat = edge_lines.materials[index]
            new_mat.setDiffuse(QColor(Qt.green))
            new_mat.setAmbient(QColor(Qt.green))
            print(f"Selected edge {index}")

    # Connect pickers (using a closure to capture 'i')
    for i, edge in enumerate(edge_lines.edges):
        for comp in edge.components():
            if isinstance(comp, Qt3DRender.QObjectPicker):
                comp.clicked.connect(lambda ev, idx=i: handle_click(idx))

    # Camera
    camera = view.camera()
    camera.lens().setPerspectiveProjection(45.0, 1.0, 0.1, 1000.0)
    camera.setPosition(QVector3D(0, 1.0, 3.0))
    camera.setViewCenter(QVector3D(0, 0, 0))

    view.setRootEntity(root_entity)

    # Orbit Controller
    cam_controller = Qt3DExtras.QOrbitCameraController(root_entity)
    cam_controller.setCamera(camera)
    cam_controller.setLinearSpeed(50.0)
    cam_controller.setLookSpeed(180.0)

    sys.exit(app.exec())

if __name__ == "__main__":
    main()
